### PR TITLE
Reconcile upstreams

### PR DESCRIPTION
**Pull request by ilackarms**
# Description

Currently if a k8s upstream is added or removed after a route is created, the changes are not reconciled by the gateway controller, as it is currently not set to resync on an Upstream change.

This PR fixes this by implementing watch for K8s Upstreams.  Verified by updating the e2e test to apply HTTPRoute in a separate step before creating the upstream
BOT NOTES: 
resolves https://github.com/solo-io/solo-projects/issues/6222

### CODE REVIEW

**Review by EItanya — DISMISSED** [review]
Needs changelog but LGTM

File: test/kubernetes/e2e/features/upstreams/suite.go [hunk 1]
```diff
@@ -36,12 +36,18 @@ func (s *testingSuite) TestConfigureBackingDestinationsWithUpstream() {
...
 		s.NoError(err, "can delete manifest")
+		err = s.testInstallation.Actions.Kubectl().DeleteFile(s.ctx, upstreamManifest)
+		s.NoError(err, "can delete manifest")
 		s.testInstallation.Assertions.EventuallyObjectsNotExist(s.ctx, proxyService, proxyDeployment)
 	})
 
 	err := s.testInstallation.Actions.Kubectl().ApplyFile(s.ctx, routeWithUpstreamManifest)
 	s.Assert().NoError(err, "can apply gloo.solo.io Upstreams manifest")
 
+	// apply the upstream manifest separately, after the route table is applied, to ensure it can be applied after the route table
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply test/kubernetes/e2e/features/upstreams/suite.go:47]: previously a single file contained both the route and the upstream. now they are split into separate files so we can verify that upstream being applied *after* the route will be picked up by the controller

File: projects/gateway2/controller/controller.go [hunk 2]
```diff
@@ -70,6 +71,7 @@ func NewBaseGatewayController(ctx context.Context, cfg GatewayConfig) error {
 		controllerBuilder.watchNamespaces,
 		controllerBuilder.watchRouteOptions,
 		controllerBuilder.watchVirtualHostOptions,
+		controllerBuilder.watchUpstreams,
```
[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/gateway2/controller/controller.go:74]: it will cause us to resync the gateway translation when an upstream is updated

**Review by EItanya — APPROVED** [review]

File: test/kubernetes/e2e/features/upstreams/suite.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply test/kubernetes/e2e/features/upstreams/suite.go:47]: Yeah that makes sense, though I think there is a naming bug here. We call ApplyFile on the routeWIthUpstreamManifest twice in a row. I think you meant to call ApplyFile on upstreamManifest here, yeah?

File: test/kubernetes/e2e/features/upstreams/inputs/route-with-upstream.yaml [hunk 3]
```diff
@@ -32,16 +32,6 @@ spec:
           urlRewrite:
             hostname: jsonplaceholder.typicode.com
 ---
-apiVersion: gloo.solo.io/v1
```
Comment (sam-heilbron) [inline test/kubernetes/e2e/features/upstreams/inputs/route-with-upstream.yaml:35]: Given that we are removing the upstream entirely from this file, the file name is misleading. Can we update it to reflect that there is no upstream?

File: projects/gateway2/controller/controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sam-heilbron) [reply projects/gateway2/controller/controller.go:74]: Yeah that makes sense. My hope here is to clarify what events could lead to that. Given that upstream discovery is disabled, are there other events in our syncer that we would expect to trigger this? Basically, if we see frequent translation occuring in a cluster, I'd like some bread crumbs to indicate what led to it

[the remark this answers is not in the document]
Comment (ilackarms) [reply projects/gateway2/controller/controller.go:74]: either discovery or a user creating/updating/deleting upstreams

File: test/kubernetes/e2e/features/upstreams/inputs/route-with-upstream.yaml (hunk 3 above)
Comment (ilackarms) [reply test/kubernetes/e2e/features/upstreams/inputs/route-with-upstream.yaml:35]: the route still points to an upstream that's why i left it

**Review by sam-heilbron — APPROVED** [review]

**Pull request merged**